src/main/java/org/jitsi/videobridge/BitrateController.java

```diff
@@ -0,0 +1,601 @@
+
+    /**
+     * The {@link PacketTransformer} that handles incoming/outgoing RTCP
+     * packets for this {@link BitrateController} instance. Internally,
+     * it delegates this responsibility to the appropriate media stream track
+     * bitrate controller ({@link SimulcastController}, etc).
+     */
+    private final PacketTransformer rtcpTransformer = new RTCPTransformer();
+
+    /**
+     * The {@link List} of endpoints that are currently being forwarded,
+     * represented by their IDs. Required for backwards compatibility with
+     * existing LastN code.
+     */
+    private List<String> forwardedEndpoints = INITIAL_EMPTY_LIST;
+
+    /**
+     * Initializes a new {@link BitrateController} instance which is to
+     * belong to a particular {@link VideoChannel}.
+     *
+     * @param dest the {@link VideoChannel} that owns this instance.
+     */
+    public BitrateController(VideoChannel dest)
+    {
+        this.dest = dest;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PacketTransformer getRTPTransformer()
+    {
+        return rtpTransformer;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PacketTransformer getRTCPTransformer()
+    {
+        return rtcpTransformer;
+    }
+
+    /**
+     *
+     * @param data
+     * @param buf
+     * @param off
+     * @param len
+     * @param source
+     * @return
+     */
+    public boolean rtpTranslatorWillWrite(
+        boolean data, byte[] buf, int off, int len, RtpChannel source)
+    {
+        Integer ssrc;
+        if (data)
+        {
+
+            long ret = RawPacket.getSSRCAsLong(buf, off, len);
+            if (ret < 0)
+            {
+                return false;
+            }
+            ssrc = (int) ret;
+        }
+        else
+        {
+            // NOTE(gp) This will need to be adjusted when we enable
+            // reduced-size RTCP.
+            long ret = RTCPHeaderUtils.getSenderSSRC(buf, off, len);
+            if (ret < 0)
+            {
+                return false;
+            }
+
+            ssrc = (int) ret;
+        }
+
+        SimulcastController subStrl = subCtrls.get(ssrc);
+
+        return subStrl != null
+            && subStrl.rtpTranslatorWillWrite(data, buf, off, len);
+    }
+
+    /**
+     *
+     * @param conferenceEndpoints
+     */
+    public void update(List<Endpoint> conferenceEndpoints)
+    {
+        // Gather the conference allocation input.
+        if (conferenceEndpoints == null)
+        {
+            conferenceEndpoints
+                = dest.getConferenceSpeechActivity().getEndpoints();
+        }
+
+        long bweBps = Long.MAX_VALUE;
+
+        // Compute the bitrate allocation.
+        EndpointBitrateAllocation[]
+            allocations = allocate(bweBps, conferenceEndpoints);
+
+        List<String> newForwardedEndpoints = new ArrayList<>();
+
+        for (int i = 0; i < allocations.length; i++)
```

Comment:
I'm favoring for loops everywhere for performance. This isn't a busy method so I'll change this to use an iterator.